### ISSUE TITLE
"Something is wrong with the file, use force reload!"

### ISSUE THREAD

**Issue by ghost** [Adamantcheese/Kuroba#978]
I get this on an image because the hash is different than expected, but nothing I do can solve it. Opening in the browser works, but is there a reason that the app can't handle these files?

**Comment by K1rakishou** [comment]
This is most likely because the server sends either corrupted image or incorrect md5 hash. Did it happen on 4chan? Would be nice to have the image and both hashes in the error message.

**Comment by ghost** [comment]
Here you go:
https://i.imgur.com/hfSZyYM.jpg
https://boards.4chan.org/v/thread/515589704

**Comment by floppyD** [comment]
Weird, it works here. Maybe your cache got corrupted?
![Screenshot_20200705-163929_Kuroba](https://user-images.githubusercontent.com/29208110/86540853-4356ca80-bede-11ea-9ac6-8d4f863d5551.png)

**Comment by K1rakishou** [comment]
Something is wrong with that image. The first time I opened it gave me one incorrect md5 hash on the second attempt it gave me a different one. Very strange.

**Comment by K1rakishou** [comment]
Yeah, something is wrong here. The actual image's md5 is F69FE89E47C91AF118B1000D752B0F02, the md5 that the 4chan's API sends is 417f1e6abd0597561f3f9468899024a4 and the one Kuroba calculates is  d710... I wonder who exactly is in the wrong here. To me seems like the server.

**Comment by Adamantcheese** [comment]
I'm also leaning on it being the server because md5sum returns that F69... value and the StringUtils.decodeBase64 method returns that same value, which means that at the very least the StringUtils function is working fine. But pretty much every single implementation of MD5 that I've tried in Android returns "d41d8cd98f00b204e9800998ecf8427e". BouncyCastle's MD5Digest, Okio's hashing sink stuff, Kotlin's ByteString stuff, hell even Perl's MD5 implementation all return that same D41... value for this file, which to me is good proof that the server messed up computing this somehow. 
I really don't know who's in the wrong here because both BouncyCastle and md5sum are reputable hashing libraries and utilities and I don't think either of them are necessarily wrong, but that doesn't make any sense. Very bizzare. Maybe it's a bug in md5sum. Either way there's not much to do.

**Comment by Adamantcheese** [comment]
I actually have an update on this because it's definitely a bug on 4chan's end.

For some images, the actual file size versus the size reported in a thread's JSON appears to be incorrect. This also affects the md5 reported versus the actual md5 of the image.

/sci/'s sticky appears to be one case of this.

For example, https://a.4cdn.org/sci/thread/5942502.json reports a file size of 8865 bytes, however curl -I https://i.4cdn.org/sci/1375476712391.jpg reports a file size of 9277 bytes, which is the same amount of bytes as what you download. This also means that the reported md5 of b7hX5iACU13d95LpyNNB2g== is incorrect; in hex that's 6fb857e62002535dddf792e9c8d341da, however the actual base64 encoded md5 of the image is NmZiODU3ZTYyMDAyNTM1ZGRkZjc5MmU5YzhkMzQxZGEK, which in hex is c172d1d06dc8a05153bfe0236743a08a.

tl:dr; literally can't fix, but I've reported it on 4chan's feedback page so maybe they'll fix it?

**Comment by Adamantcheese** [comment]
And another example is the photo in /qa/'s sticky. Somehow the second image in the sticky thread is fine.

**Comment by K1rakishou** [comment]
I think the best solution here would be to remove the md5 check (compareFileHashes) in ChunkMerger. I initially added it as a test to make sure the file is stored on the disk correctly and it turns out there were no bugs related to that so it should be fine. Or maybe make it a setting.

**Comment by Adamantcheese** [comment]
Removing the md5 check would still result in corrupt files because the file being downloaded is (in the example above) 8865 bytes instead of 9277 bytes. A better solution would be to just never trust any site's reported file size/hash and always send a HEAD request.

**Comment by K1rakishou** [comment]
>Removing the md5 check would still result in corrupt files because the file being downloaded is (in the example above) 8865 bytes instead of 9277 bytes.

Oh yes, you are correct.

>A better solution would be to just never trust any site's reported file size/hash and always send a HEAD request.

It will make images loading slower especially on 4chan. There is also a timeout I use when sending HEAD requests, 1 second. If the request didn't complete in 1 second then it's canceled. So if you remove that too then image loading will get even slower.

**Comment by Adamantcheese** [comment]
Eh, it happens rarely enough on 4chan (just these few strangely specific instances?) that I think it's fine to leave it as-is then. Really strange though how those post images seem to be bad.

**Comment by Adamantcheese** [comment]
Some additional info: a bunch of really old posts on /po/ seem to have this issue as well. Guessing at some point in history something went wrong on 4chan's end. I'm just going to close this one out for good here with that knowledge.
